modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -2765,3 +2765,9 @@ def finalize(self):
         that were used to build it.
         """
         self._partition_mgr_cls.finalize(self._partitions)
+
+    def num_chunks(self):
```

Comment:
If we're going to implement the protocol, I prefer to do it at this level. Each implementation under the query compiler may want a different implementation.

Reply:
Definitely. So far, it is just a copy of the protocol implementation for pandas.

Reply:
To make sure, we are on the same page. Is your preference that we should do general things at the API layer and put implementation-specific details at a concrete dataframe layer?

Reply:
I prefer that we implement the exchange protocol at out internal dataframe implementation for a couple of reasons:

1.) Other implementations (e.g. omnisci) may want to implement a separate, more efficient interchange protocol implementation
2.) We may want to interchange between Modin implementations using this protocol (e.g. omnisci -> pandas on ray), and it is harder to do that at upper layers.

Reply:
Yes, I agree. That makes sense.

Reply:
Look at the changes. That is exactly what you mean? I am not sure if a conflict of import dependencies may happen in that case.

Reply:
@YarShev yes, this is what I mean. Does it make sense to you?

Reply:
I have a couple of questions/comments on your thoughts.

> 1.) Other implementations (e.g. omnisci) may want to implement a separate, more efficient interchange protocol implementation.

We can still implement the protocol for omnisci/pandas_on_ray/dask (through modin.pandas.df -> qc -> internal_modin_df) if we put the protocol at the upper layer. What is the difference between that if we put the protocol at lower layers?

> 2.) We may want to interchange between Modin implementations using this protocol (e.g. omnisci -> pandas on ray), and it is harder to do that at upper layers.

I don't quite get why it is harder to interchange data between Modin implementations for omnisci <-> pandas_on_ray (between others too) at the upper layer. Could you elaborate?

I think the implementation would be simpler at the upper layer because the upper layer DataFrame has a lot of functionality that is not present in the core Modin Dataframe (for example, iloc, loc, df[index1:index2], isna, and others). We by far can implement that via mask, map, etc. over the core Modin Dataframe but we have to use more complex constructions. Also, the core Modin Dataframe should expose an interface for the dataframe algebra (some utilities for the algebra too) but in addition we impose it to have the protocol interface. Is that okay?

I am not saying the upper layer would be better to put the protocol implementation but would like we reach agreement on a more appropriate place for it.